xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+        """
+        self.add(callable)
+        return callable
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+
+        The caller may send() new positional arguments (eg, to implement
+        modifying semantics). Keyword arguments cannot be modified this way.
+        """
+        for handler in self:
+            newargs = yield handler(*pargs, **kwargs)
```

Comment:
>  I would call this deterministic but chaotic, and not easily reasoned about.

Exactly. I feel that this needs to be easily reasoned about. It is a public interface.

> Which makes me think that for the sake of future sanity, any kind of event semantics that depend on ordering should be avoided.

Yes, I agree. These kinds of things should be avoided if at all possible.  However, avoided doesn't mean forbidden.  There may even be cases in the core that may be most easily implemented in this way.  I don't think we need to implement a full DAG dependency solver for this, but we also don't want to explicitly disallow any ad hoc dependencies either.